Caret behavior by long tap in cupertino textfield

mazunin-v-jb opened the pull request:
## Proposed Changes

Added native behavior for long tap and drag in Cupertino text field. Unlike Android, long tap and drag gesture doesn't enter selection mode, It just moves the cursor.

## Testing

Test: Open test app, go Components -> Text fields -> Almost fullscreen, tap on the textfield to focus it, then try to long tap in somewhere else and drag.

## Issues Fixed

- https://youtrack.jetbrains.com/issue/COMPOSE-449/iOS-caret-movement-by-long-tap-in-textfields
- https://youtrack.jetbrains.com/issue/COMPOSE-448/iOS-separate-long-tap-logic-and-drag-selection-logic-in-textfields (it is related to the issue above)

## Google CLA
You need to sign the Google Contributor’s License Agreement at https://cla.developers.google.com/.
This is needed since we synchronise most of the code with Google’s AOSP repository. Signing this agreement allows us to synchronise code from your Pull Requests as well.

igordmn committed:
Fix Accessibility. Crash on text hover on Windows (#852)

Fix Accessibility. Crash on text hover on Windows

Fixes https://github.com/JetBrains/compose-multiplatform/issues/3742

The fix is cherry-picked from
https://github.com/androidx/androidx/commit/7a7f57c49041d64f4c88b28686979f325cc6fd7c
to jb-main
(it contains other fix as well)

Because we always return `false`, we [have textLayoutResult =
null](https://github.com/JetBrains/compose-multiplatform-core/blob/7f72ab8719252acaafc753a9aa28d2940ea995df/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/ComposeAccessible.kt#L116)

It isn't reproducible on macOs because it doesn't call
`ComposeAccessibleText.getCharacterBounds`

# Test
1. Modified AccessibilityTest (it fails before the fix on Windows/macOs)
2. VoiceOver pronounces the hovered Text on macOs, Windows

MatkovIvan committed:
Add unit test for #847 regression (#856)

## Proposed Changes

- Make `Popup`/`Dialog` insets overridable (internally)
- Move platform insets specifics into `PlatformInsets` file
- Add test to cover #847 regression

## Testing

Test: run `PopupTest.popupBoundsWithPlatformInsets` test

---------

Co-authored-by: Igor Demin <igordmn@users.noreply.github.com>

elijah-semyonov committed:
iOS UIKit interop regression fix (#857)

elijah-semyonov committed:
Fix long text field overscroll effect not clipped correctly (#859)

## Proposed Changes

Make CupertinoOverscrollEffect add `clip` Modifier before `offset` if
requested by a using widget.
Some widgets (like LazyList) do it internally, so adding it
unconditionally will be redundant.

## Testing

Test: reported case should clip correctly now.

## Before


https://github.com/JetBrains/compose-multiplatform-core/assets/4167681/2158106e-f782-4824-b46c-c576859c6d0e

## After


https://github.com/JetBrains/compose-multiplatform-core/assets/4167681/da8ebf8a-4106-4d19-9bac-49963eae3977

<img width="413" alt="Screenshot 2023-10-04 at 11 48 36"
src="https://github.com/JetBrains/compose-multiplatform-core/assets/4167681/eb0523b3-c412-4bf7-bbf5-6b5a94cc45ac">

## API change

`CupertinoOverscrollEffect` (uikit source set) now has an additional
argument `val applyClip: Boolean` in the constructor.
`CupertionOverscrollEffect` is marked as `@ExperimentalFoundationApi`
now.

---------

Co-authored-by: dima.avdeev <dima.avdeev@jetbrains.com>

m-sasha committed:
Fix accessibility not reporting changes (#842)

m-sasha committed:
Return `null` from ComposeAccessible.getAccessibleContext when the node has been removed (#865)

m-sasha committed:
Report semantic `ProgressBarRangeInfo` changes. (#862)

m-sasha committed:
Avoid calling Window.setVisible(true) on every `AwtWindow` recomposition (#863)

eymar committed:
iOS: Fix first screen is recomposed twice (#861)

Fixes: https://github.com/JetBrains/compose-multiplatform/issues/3778

Implemented according to what @elijah-semyonov suggested:

> We need to defer the first recomposition until `didMoveToWindow` was
called with non-nil window. That's when a correct density scale is
available.

---------

Co-authored-by: dima.avdeev <dima.avdeev@jetbrains.com>

dima-avdeev-jb committed:
iOS. Fixed bug with selection handle (#869)

Previously SelectionHandle looks like this on BasicTextField:

It was because of Popup padding
Not, it was fixed:

dima-avdeev-jb committed:
reproduce and reminder about iOS SelectionContainer bug (#867)

Made reproducer of a bug
https://github.com/JetBrains/compose-multiplatform/issues/3718
This bug should be fixed after merging newer version of androidx-main to
jb-main branch.
Added TODO comment to check and close related issues after that.

igordmn committed:
Don't send synthetic Move events before Press/Release for touch (#870)

These kind of synthetics were added to fix missed Enter/Exit events:

Box1 | Box2

If we have this chain of native events:
```
Move on Box1
Press on Box1
Release on Box2
Move on Box2
```
Without synthetic events, Compose receives:
```
Enter on Box1
Press on Box1
Release on Box1 (because Box1 takes ownership of the all events)
Enter on Box2
```
With synthetic events:
```
Enter on Box1
Press on Box1
Exit on Box1
Release on Box1
Enter on Box2
```

Touch doesn't have Enter/Exit, so we don't have to send them.
Furthermore, such kind of synthetics can bring issues when calculating
the end fling velocity (the end position should be excluded from the
velocity calculation on iOS). This isn't an issue for mouse.

## Testing

- run existing tests
- run iOS demo and check that click, and scrolling works

## Issues Fixed


https://youtrack.jetbrains.com/issue/COMPOSE-474/Fix-synthetic-events-before-end-event-with-different-position

elijah-semyonov committed:
Ignore unpressed events during velocity calculation on iOS (#848)

dima-avdeev-jb committed:
COMPOSE-499 Fix crash with Asian languages in TextField. (#872)

https://github.com/JetBrains/compose-multiplatform/issues/3811 
 - Added basic implementations instead of TODOs.
- Add additional checks to position argument, because it may be nullable
in some methods.

igordmn committed:
Update Skiko to 0.7.84, fix failed tests after updating coroutines (#871)

- Update Skiko to 0.7.84
([Diff](https://github.com/JetBrains/skiko/compare/v0.7.80...v0.7.84))
- Because it updates Coroutines to 1.7.3, now we have failing tests,
they are migrated to the new coroutine version

Having 1.7.3 in dependencies is okay, as it depends on [Kotlin
1.8.20](https://github.com/Kotlin/kotlinx.coroutines/blob/master/CHANGES_UP_TO_1.7.md#version-170)
(Compose 1.5 depends on 1.8.21). Besides that, [Jetpack Compose 1.6
already depends on
1.7.1](https://github.com/androidx/androidx/blob/e502da69dcec7dc8662241de58f34399e7b0abe8/gradle/libs.versions.toml#L47)

MatkovIvan committed:
Remove unnecessary state from composition local (#873)

## Proposed Changes

Changing API marked as internal (`@InternalComposeApi`)

```diff
-androidx.compose.ui.uikit.LocalInterfaceOrientationState
+androidx.compose.ui.uikit.LocalInterfaceOrientation
```
```diff
-androidx.compose.ui.uikit.LocalKeyboardOverlapHeightState
+androidx.compose.ui.uikit.LocalKeyboardOverlapHeight
```


## Testing

Test: run test app on iOS

Walingar committed:
Support a11y navigation for Swing interop (#879)

Enable a11y navigation, so focus change inside Compose component will be
properly handled by a11y support

Walingar committed:
Fix NPE for getComponentAfter/Before in ComposePanel (#878)

These methods are called on remove(ComposePanel.desktop.kt:121), and
since NPE can be thrown bridge property doesn't become null
(ComposePanel.desktop.kt:122)

Related issue: https://github.com/JetBrains/jewel/issues/182

igordmn committed:
Take into account `enabled` in `scrollable` for mouse input (#880)

Fixes https://github.com/JetBrains/compose-multiplatform/issues/3816

## Testing
1. The new test
2. Manual testing of `Modifier.verticalScroll(enabled = false`

eymar committed:
Remove @ExportObjCClass from SkikoUIView (#883)

We don't use this class from ObjC. There the annotation is redundant

This change also fixes the warning "Class SkikoUIView has multiple
implementations. Which one will be used is undefined."

igordmn committed:
Update Skiko to 0.7.85 (#886)

- Update Skiko to 0.7.85
([Diff](https://github.com/JetBrains/skiko/compare/v0.7.84...v0.7.85))

igordmn committed:
Desktop. Don't use AccessibleRole.PANEL for root/unknown nodes (fixes issues on Windows) (#885)

This fixes issues on Windows:

1. Windows pronounced "Panel unavailable" at start
2. When we focus on Switch or Box(Modifier.clickable), there was
"Panel0" pronounced. Now it says nothing.
3. After merging
https://github.com/JetBrains/compose-multiplatform-core/pull/881, we
refocus the main component. This leads to pronouncing the window title,
now it says nothing, as the accessible root node is unknown

on macOs it changes one thing:
1. When we focus on Switch or Box(Modifier.clickable), now it prounonces
that we in scroll area (if we in scroll area). It also pronounces it for
other compnents like Button/Text

## Testing
Manually with accessibility enabled in run1 (Windows, macOs)

igordmn committed:
Desktop. Fix input methods on JBR, disable input methods when we lose focus (#881)

### Rerequest focus on main component when we need to type using input
methods
Fixes https://github.com/JetBrains/compose-multiplatform/issues/2628

The issue was because of 2 things:
- we used a hack to force a focused event
(`component.inputContext.dispatchEvent(focusGainedEvent)`
- JBR added optimization to ignore focus on the same element (thanks
@AiHMin for investigation
[here](https://github.com/JetBrains/compose-multiplatform/issues/2628#issuecomment-1704562796)).
In Compose we have only one element.
Because optimization is correct, and the hack depends on the internals,
it isn't right to fix it in JBR, we should fix it in Compose.
Furthermore, even without JBR changes, this hack didn't complete work -
we can't for example use it for disabling input methods (see the next
point).

In this PR we also use a hack unfortenutely - we refocus the root
component, focusing on invisible component first. That leads to another
issue with acccessibility, but we fix it
[here](https://github.com/JetBrains/compose-multiplatform-core/pull/885)).

A proper fix should be switching to native code, or making an API in JBR
(but other vendors still be unsupported).

### Don't show input methods popup if there is no focused TextField 

Previously we showed a popup, even if there are no focused textfield:

![image](https://github.com/JetBrains/compose-multiplatform-core/assets/5963351/82e3543f-11f4-4013-8da5-d782b824ed2c)
Now we don't show it if we isn't in a textfield.
P.S. only on Windows/Linux for now, on macOs Swing seems has [a
bug](https://github.com/JetBrains/compose-multiplatform/issues/3839)

## Testing
Tested manually on:
1. Windows, Chinese/Korean/Japanese, OpenJDK/JBR 17, Accessibility
enabled/disabled, ComposeWindow/ComposePanel
2. macOs, Chinese, OpenJDK/JBR 17, Accessibility enabled/disabled
4. Linux, Chinese layout, OpenJDK 17

terrakok committed:
Synchronize IME insets with iOS keyboard (#875)

Proper support of the keyboard animation

---------

Co-authored-by: dima.avdeev <dima.avdeev@jetbrains.com>

MatkovIvan committed:
[mpp test app] Check nav stack size to avoid crash on double click (#887)

```kt
Uncaught Kotlin exception: kotlin.NoSuchElementException: List is empty.
    at 0   ComposeDemo                         0x10047954b        kfun:kotlin.Throwable#<init>(kotlin.String?){} + 123 
    at 1   ComposeDemo                         0x100471bcf        kfun:kotlin.Exception#<init>(kotlin.String?){} + 119 
    at 2   ComposeDemo                         0x100472013        kfun:kotlin.RuntimeException#<init>(kotlin.String?){} + 119 
    at 3   ComposeDemo                         0x100471e9b        kfun:kotlin.NoSuchElementException#<init>(kotlin.String?){} + 119 
    at 4   ComposeDemo                         0x1004be99f        kfun:kotlin.collections#last__at__kotlin.collections.List<0:0>(){0§<kotlin.Any?>}0:0 + 335 
    at 5   ComposeDemo                         0x10166590b        kfun:androidx.compose.mpp.demo.App#Content(androidx.compose.runtime.Composer?;kotlin.Int){} + 1587 
    at 6   ComposeDemo                         0x1018afa17        kfun:androidx.compose.mpp.demo#IosDemo(kotlin.String;androidx.compose.runtime.Composer?;kotlin.Int){} + 3455 
    at 7   ComposeDemo                         0x1018b0277        kfun:androidx.compose.mpp.demo.main$lambda$0#internal + 615 
    at 8   ComposeDemo                         0x1018b04c7        kfun:androidx.compose.mpp.demo.$main$lambda$0$FUNCTION_REFERENCE$0.invoke#internal + 139 
    at 9   ComposeDemo                         0x1018b0633        kfun:androidx.compose.mpp.demo.$main$lambda$0$FUNCTION_REFERENCE$0.$<bridge-UNNNB>invoke(androidx.compose.runtime.Composer;kotlin.Int){}#internal + 183 
    at 10  ComposeDemo                         0x1007bb39b        kfun:androidx.compose.runtime.internal.ComposableLambdaImpl#invoke(androidx.compose.runtime.Composer;kotlin.Int){}kotlin.Any? + 875
```

FaroukAbichou committed:
Fixed Typo in Application.desktop.kt (#888)

## Proposed Changes

  - Fixed typo in Application.desktop.kt

alexzhirkevich committed:
Fix format with skeleton (#877)

## Proposed Changes

Fixed  https://kotlinlang.slack.com/archives/CJLTWPH7S/p1697506072139909

 Can you please run tests for uikit and check datepicker on device with
`NSTimeZone.setDefaultTimeZone(NSTimeZone.timeZoneWithName("GMT-5")!!)`

## Testing

Test: new unit tests

dima-avdeev-jb committed:
Reproduce iOS bug with keyboardType change (#895)

Reproduce iOS bug
https://github.com/JetBrains/compose-multiplatform/issues/3885

elijah-semyonov committed:
Make Skia context usage a critical section on iOS. (#896)

## Proposed Changes

Since there are no reliable repros, it's a speculative fix for [the
crash](https://github.com/JetBrains/compose-multiplatform/issues/3862)
on iOS.
It's based on an assumption that the case for the crash is caused by
render command encoding in a separate thread being performed after (or
in parallel) with the context disposal on the main thread which leads to
incorrect state inside Skia.

## Testing

Test: see if issues persists.

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-multiplatform/issues/3862

## Note
Skia is supposed to handle internal resources based on reference
counting and assumed scenario shouldn't lead to the crash, since the
context should be indirectly retained by the moment encoding starts.
Revert if issue persists because that logic would be redundant.

JakeWharton committed:
Remove unused `Pair` creation (#892)

Its instantiation was never used.

JakeWharton committed:
Unify runtime-saveable's targets with the runtime (#894)

It supports more targets than the general compose multiplatform
libraries, and the two should be unified.

Test: gw :compose:runtime:runtime-saveable:watchosArm64Binaries
:compose:runtime:runtime-saveable:watchosArm32Binaries
:compose:runtime:runtime-saveable:watchosX86Binaries
:compose:runtime:runtime-saveable:watchosX64Binaries
:compose:runtime:runtime-saveable:watchosSimulatorArm64Binaries
:compose:runtime:runtime-saveable:tvosArm64Binaries
:compose:runtime:runtime-saveable:tvosX64Binaries
:compose:runtime:runtime-saveable:tvosSimulatorArm64Binaries
:compose:runtime:runtime-saveable:mingwX64Binaries
:compose:runtime:runtime-saveable:linuxX64Binaries

Closes https://github.com/JetBrains/compose-multiplatform/issues/3899

sebkur committed:
Use larger icon size for Window's IconImage (#890)

This avoids blurry icons on window switchers (Alt+Tab) on Windows and
Linux.

## Proposed Changes

- change icon image size for windows to a larger base image 192px
instead of 32px, otherwise we end up with blurry icons (see
https://github.com/JetBrains/compose-multiplatform/issues/1838)

## Testing

Test: I have not tested this change directly, but am using this size on
other apps to set the size and this seems to work well on the systems I
tried on.

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-multiplatform/issues/1838

## Google CLA
You need to sign the Google Contributor’s License Agreement at
https://cla.developers.google.com/.
This is needed since we synchronise most of the code with Google’s AOSP
repository. Signing this agreement allows us to synchronise code from
your Pull Requests as well.

MatkovIvan committed:
Update skiko to 0.7.87 (#899)

## Proposed Changes

- Update skiko to `0.7.87`

## Issues Fixed

Fixes https://github.com/JetBrains/compose-multiplatform/issues/1888
<img width="310" alt="image"
src="https://github.com/JetBrains/compose-multiplatform-core/assets/1836384/dd28cc81-1d5d-4119-b9fa-0e1e4e28f794">

---------

Co-authored-by: Igor Demin <igor.demin@jetbrains.com>

MatkovIvan committed:
[Tests] Use `ComposeRootRegistry` instead of scene tree (#893)

## Proposed Changes

It's a part of #891

- Add callbacks to `SkiaRootForTest` to collect references in
`ComposeRootRegistry` (original Google's approach like in the Android
source set)
- Remove "child" scenes introduced in #697. Currently it's handled in
`ComposeRootRegistry` (tests-only)
- Deprecate `ComposeScene.roots` that was never intended to be public
- Address TODO in `isInScreenBounds()`: `contentSize` instead of
`containerSize` where used

## Testing

Test: CI should pass, everything should work as before

MatkovIvan committed:
Add `ComposeScene.semanticsOwner` (#900)

## Proposed Changes

```diff
+ComposeScene.semanticsOwner
```

## Issues Fixed

Fixes
https://github.com/JetBrains/compose-multiplatform-core/pull/893#discussion_r1388706236

MatkovIvan committed:
Add a way to use fonts installed on the system (#898)

## Proposed Changes

- Add `SystemFont` class to be able to use fonts installed on the
system.

## API Change

```diff
+androidx.compose.ui.text.platform.SystemFont
```

## Testing

Usage looks like this:
```kt
FontFamily(SystemFont("Menlo")),
FontFamily(SystemFont("Times New Roman", FontWeight.Bold)),
FontFamily(SystemFont("Webdings")),
```
<img width="1003" alt="Screenshot 2023-11-09 at 12 18 02"
src="https://github.com/JetBrains/compose-multiplatform-core/assets/1836384/0c36d4aa-6bc9-47e2-94dd-edb09fd242ac">

MatkovIvan committed:
Add support `LineHeightStyle.Trim` (#897)

## Proposed Changes

- Set `ParagraphStyle.heightMode` based on `LineHeightStyle.Trim` value
- Align default behavior with Android
- Avoid using `StrutStyle` - it doesn't allow the line height trimming.
Set `height` via `TextStyle` instead
- Cache and post-process `lineMetrics`
- Port tests from an Android source set

## Behaviour change

In case of larger `lineHeight`, both paddings are trimmed by default to
match the Android behaviour.

## Testing

Test: run tests from `DesktopParagraphIntegrationLineHeightStyleTest`

```kt
Row(horizontalArrangement = Arrangement.spacedBy(5.dp)) {
    for (lineHeightStyle in listOf(
        null,
        LineHeightStyle(LineHeightStyle.Alignment.Proportional, LineHeightStyle.Trim.Both),
        LineHeightStyle(LineHeightStyle.Alignment.Proportional, LineHeightStyle.Trim.FirstLineTop),
        LineHeightStyle(LineHeightStyle.Alignment.Proportional, LineHeightStyle.Trim.LastLineBottom),
        LineHeightStyle(LineHeightStyle.Alignment.Proportional, LineHeightStyle.Trim.None),
    )) {
        Text("Line 1\nLine 2",
            modifier = Modifier.background(Color.Gray),
            style = TextStyle(
                fontSize = 18.sp,
                lineHeight = 50.sp,
                lineHeightStyle = lineHeightStyle
            )
        )
    }
}
```

Before | After
--- | ---
<img width="285" alt="Screenshot 2023-11-08 at 13 38 08"
src="https://github.com/JetBrains/compose-multiplatform-core/assets/1836384/2c7fbddc-dac9-408e-ad53-cedc361fe777">
| <img width="285" alt="Screenshot 2023-11-08 at 13 37 45"
src="https://github.com/JetBrains/compose-multiplatform-core/assets/1836384/4fe4d7c9-a414-4338-885f-6701e6daecf1">



## Issues Fixed

Fixes (partially)
https://github.com/JetBrains/compose-multiplatform/issues/2602
Fixes https://github.com/JetBrains/compose-multiplatform/issues/3866

MatkovIvan committed:
Copy `SystemFont` handling to JS (#901)

## Proposed Changes

- Fixes JS build after #898
- Does **NOT** fix fonts in JS
(https://github.com/JetBrains/compose-multiplatform/issues/3915), so
it's compilation only fix.

dima-avdeev-jb committed:
Update Skiko 0.7.88 (#904)

Walingar committed:
Provide a way to calculate ComposeScene preferredSize that is needed for Swing (#884)

---------

Co-authored-by: Igor Demin <igordmn@users.noreply.github.com>

elijah-semyonov committed:
Disable encoding on separate thread for iOS (#907)

## Proposed Changes

Using Skia context in multiple threads simultaneously leads to
occasional unreproducable crashes on users' side.
Roll back experimental #896. Disable the path for encoding rendering
commands on a separate thread until the scenario is resolved and
underlying issue is fixed.

## Testing

Test: N/A

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-multiplatform/issues/3862

elijah-semyonov committed:
UIViewController interop (#882)

## Proposed Changes

Add API for embedding UIViewControllers for cases, where UIView-based
API is not available (like SwiftUI integration via UIHostingController,
UITabBarController, UINavigationController, etc.).

## Testing

Test: check "SwiftUI interop example" in Demo.

<img width="424" alt="Screenshot 2023-10-19 at 15 18 16"
src="https://github.com/JetBrains/compose-multiplatform-core/assets/4167681/0e7e23b5-687f-40a0-a574-cff48e951acb">

m-sasha committed:
Move the call to scene.render in tests into the UI thread (#910)

To avoid race conditions with GlobalSnapshotManager and effects
dispatcher, move the call to scene.render in tests into the UI thread.

eymar committed:
Cherry-pick `Fix missing return statement in Composable functions when it's the last statement and returns a null const` https://android-review.googlesource.com/c/platform/frameworks/support/+/2835716

sanao1006 committed:
docs: Update compiler-metrics.md (#909)

## Proposed Changes
- Add `/` to gradlew command because it was missing.
## Testing
Test: N/A

dima-avdeev-jb committed:
Font with lazy get data in skikoMain (#906)

We just added new function Font with lazy getData lambda.
We don't need changes in Skiko library.

mazunin-v-jb committed:
Implemented long press logic in cupertino textfield

mazunin-v-jb committed:
Implemented long press logic in cupertino textfield

mazunin-v-jb commented:
Should be merged only after https://github.com/JetBrains/compose-multiplatform-core/pull/858

eymar committed:
Make OEL publication compatible with kotlin 1.9.0 and newer (#841)

Kotlin >= 1.9.0 had an implementation change which made our OEL
publication crash. We were using reflection to access some APIs.
This PR updates `publishAndroidxReference` function making it compatible
with current kotlin version a newer ones.

MatkovIvan committed:
Refactor `ComposeScene` (#908)

## Proposed Changes

It's the evolution of #891

Compose scene:
- Deprecate the existing API and introduce an internal interface instead
- Remove key listeners from `setContent`
- `constraints` -> `size`
- Add `lastKnownPointerPosition` to public interface 
- `CombinedComposeScene` as a compatible version of scene
- Remove recently added `semanticsOwner`
- `ComposeScene.Pointer` -> `ComposeScenePointer`

Other changes:
- Rename `SkiaBasedOwner` to `RootNodeOwner`. Also, it uses aggregation
instead of implementing a few interfaces.
- `SkikoComposeUiTest` overrides `PlatformContext.RootForTestListener`
now. It solves TODOs about `WindowInfo` in tests.
- `SkiaRootForTest` is renamed to `PlatformRootForTest`
- `onRootCreatedCallback`/`onRootDisposedCallback` is now part of
`PlatformContext`
- `PlatformRootForTest` exposes real `visibleBounds` instead of just
window size information.
- `PlatformRootForTest` exposes send pointer input methods instead of
`scene` reference.
- `InternalComposeUiApi` is opted in inside the same module.
- `Platform` -> `PlatfromContext`
- `dialogScrimBlendMode` -> `isWindowTransparent`
- `AccessibilityController`: Remove interface from `skikoMain` source
set
- `AccessibilityController`: `syncLoop` triggered right from
desktop-only part
- Replaced `Platform.accessibilityController(SemanticsOwner)` to
`semanticsOwnerListener`
- Fixes an invalid test case: out of screen Popup cannot receive input
now
- Removed unnecessary expect/actual `createSkiaLayer`
- Moves "Deprecated" `DefaultViewConfiguration` to test where there is a
single usage of it.

## API Changes

See `compose/ui/ui/api/desktop/ui.api` file

## Testing

Test: run existing tests.

dima-avdeev-jb committed:
remove redundant expect modifier for ComposeWindow class (#866)

We don't need expect / actual ComposeWindow
I mentioned this in Architecture of iOS meeting

eymar committed:
Update skiko to 0.7.89 (#916)

Updating skiko version.

I'd like the next dev build to include a change from skiko with a new
default font for web targets.

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -48,7 +49,7 @@ internal fun Modifier.cupertinoTextFieldPointer(
 ): Modifier = if (enabled) {
     if (isInTouchMode) {
-        val selectionModifier = getSelectionModifier(manager)
+        val selectionModifier = getLongPressHandlerModifier(state, offsetMapping)
```
```suggestion
        val longPressHandlerModifier = getLongPressHandlerModifier(state, offsetMapping)
```

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -132,24 +133,86 @@ private fun getTapHandlerModifier(
+    val currentOffsetMapping by rememberUpdatedState(offsetMapping)
+    var dragTotalDistance = remember { Offset.Zero }
+    var dragBeginOffset = remember { Offset.Zero }
+    var currentDragPosition: Offset?
```
We shouldn't store variables without remembering them. Any recomposition will erase this value.

Move it as a field of `longTapActionsObserver`

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -132,24 +133,86 @@ private fun getTapHandlerModifier(
+    var dragBeginOffset = remember { Offset.Zero }
+    var currentDragPosition: Offset?
+
+    val longTapActionsObserver = object : TextDragObserver {
```
```suggestion
    val longTapActionsObserver = remember {
        object : TextDragObserver {
    }
```

We shouldn't create a new object every recomposition and pass it to `pointerInput(Unit)` coroutine. We should create it once and `remember`.

If we won't do that, the coroutine will see the object only from the first recomposition. We will have a memory leak and input won't work, as it will manipulate the lost object.

igordmn replied:
As an alternative to `remember`, we can move it to the coroutine itself:
```
return Modifier.pointerInput(Unit) {
    val longTapActionsObserver = object : TextDragObserver {
        ...
    }
    detectDragGesturesAfterLongPress(
```
This looks better

mazunin-v-jb replied:
Did I understand correctly that this in this particular case will have the similar effect like remember `TextDragObserver` object outside of the lambda's scope of `Modifier.pointerInput(Unit)`, but more lazy?
If there was more than one usage of `TextDragObserver` object, we would prefer to use remember instead?

igordmn replied:
> similar effect like remember

Yes, the coroutine is only created once, and suspends on `detectDragGesturesAfterLongPress`

> but more lazy?

It is created at another time, yes, but this only give us a better code (related pieces defined in one place.)

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -132,24 +133,86 @@ private fun getTapHandlerModifier(
+): Modifier {
+    val currentState by rememberUpdatedState(state)
+    val currentOffsetMapping by rememberUpdatedState(offsetMapping)
+    var dragTotalDistance = remember { Offset.Zero }
```
We can move `dragTotalDistance` and `dragBeginOffset` as fields of anonymous object `longTapActionsObserver`

mazunin-v-jb replied:
Didn't get used to that it's possible

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -132,24 +133,86 @@ private fun getTapHandlerModifier(
+
+        override fun onDrag(delta: Offset) {
+            dragTotalDistance += delta
+            state.layoutResult?.let { layoutResult ->
```
```suggestion
            currentState.layoutResult?.let { layoutResult ->
```

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -132,24 +133,86 @@ private fun getTapHandlerModifier(
+        override fun onDrag(delta: Offset) {
+            dragTotalDistance += delta
+            state.layoutResult?.let { layoutResult ->
+                currentDragPosition = dragBeginOffset + dragTotalDistance
```
```suggestion
                val currentDragPosition = dragBeginOffset + dragTotalDistance
```
We don't use `currentDragPosition` anywhere, except this function.

igordmn commented on compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/text/CupertionTextFieldPointerModifier.skiko.kt:
```diff
@@ -132,24 +133,86 @@ private fun getTapHandlerModifier(
+        }
+
+        override fun onStop() {
+            currentState.showFloatingToolbar = true
```
Now we have a concurrent ownership with `TextFieldSelectionManager` of writing to `showFloatingToolbar`. `TextFieldSelectionManager` can change it to `false`, and we'll reset it to `true`.

If we need to disable the toolbar during drag, we need to introduce 2 variables, and write a combined value to `showFloatingToolbar`

mazunin-v-jb replied:
I did this similarly to other `TextDragObserver` objects. I tested it more thoroughly and decided that we don't need to handle the state of `showFloatingToolbar` here at all.
Deleted this from here.

mazunin-v-jb committed:
review fixes: renaming variables, memory handling, remove redunant code

mazunin-v-jb committed:
review fixes: renaming variables, memory handling, remove redunant code

eymar committed:
Use an alternative implementation of Image.toBitmap for web target (#917)

`Bitmap.makeFromImage` is faster than `canvas.drawImage` on web:

With `canvas.drawImage`:
First call: 44ms; Second call: 14ms; Third call: 16ms; (for 256x256 PNG
image)


With `Bitmap.makeFromImage`:
First call: 9.92ms; Second call: 1.88ms; Third call: 1.63ms; (for
256x256 PNG image)

Please see a comment in the code for more details.

---------

Co-authored-by: Oleksandr Karpovich <oleksandr.karpovich@jetbrains.com>
Co-authored-by: Igor Demin <igordmn@users.noreply.github.com>

m-sasha committed:
Fix positioning of interop views. (#922)

alexzhirkevich committed:
Fix datepicker empty row (#921)

m-sasha committed:
Use LayoutCoordinates.positionInWindow() instead of LayoutCoordinates.localToWindow(Offset.Zero) (#923)

MatkovIvan committed:
Avoid unnecessary composition recreation in `Dialog` (#925)

## Proposed Changes

- Fixes regression after #908
- Repeated call of `setContent` recreates recomposition that breaks the
state.

## Testing

```kt
fun main() = singleWindowApplication {
    var text by remember { mutableStateOf("") }
    Dialog(
        onDismissRequest = {  },
    ) {
        Box(modifier = Modifier.background(Color.White).size(400.dp, 300.dp)) {
            OutlinedTextField(
                value = text,
                onValueChange = { text = it },
                modifier = Modifier.padding(32.dp)
            )
        }
    }
}
```

eymar committed:
Add K/Wasm support (#914)

This PR:
- updates kotlin to 1.9.21
- removes watchosX86() target - it's not available since kotlin 1.9.20

---------

Co-authored-by: Oleksandr Karpovich <oleksandr.karpovich@jetbrains.com>

mazunin-v-jb committed:
Caret behavior by single tap in Cupertino textfields (#858)

## Proposed Changes

- Split behavior for focused and unfocused state, like in native
applications.
  - Added behavior logic for focused state of text field.
- Forwarded callback for showing context menu actions by tap on the
caret.

## Testing

Test: Run App, open Components - TextField - Almost Fullscreen. Try
cases described in the documentation of
`cupertinoSetCursorOffsetFocused() `

## Issues Fixed


https://youtrack.jetbrains.com/issue/COMPOSE-445/iOS-adjust-caret-behavior-in-textfields-for-single-tap


## Google CLA
You need to sign the Google Contributor’s License Agreement at
https://cla.developers.google.com/.
This is needed since we synchronise most of the code with Google’s AOSP
repository. Signing this agreement allows us to synchronise code from
your Pull Requests as well.

mazunin-v-jb committed:
Merge branch 'jb-main' into v.mazunin/dev/long-tap-native-behavior-in-cupertino-textfield
